runtime/lua/vim/pack.lua

```diff
@@ -0,0 +1,653 @@
+    opt = pkg.opt,
+    pin = pkg.pin,
+    status = uv.fs_stat(dir) and M.status.INSTALLED or M.status.TO_INSTALL,
+    url = url,
+  }
+end
+
+--- @param pkg vim.pack.Package
+--- @param counter function
+local function remove(pkg, counter)
+  local ok = pcall(vim.fs.rm, pkg.dir, { recursive = true })
+  counter(pkg.name, Messages.remove, ok and 'ok' or 'err')
+  if not ok then
+    return
+  end
+  pkg.status = M.status.REMOVED
+  Packages[pkg.name] = pkg
+end
+
+--- @nodoc
+--- @class pack.Conflict
+--- @field prev vim.pack.Package
+--- @field curr vim.pack.Package
+---
+--- @return pack.Conflict[]
+local function calculate_conflicts()
+  local conflicts = {}
+  for name, lock in pairs(Lock) do
+    local pkg = Packages[name]
+    if pkg and Filter.not_removed(lock) and (lock.branch ~= pkg.branch or lock.url ~= pkg.url) then
+      table.insert(conflicts, { prev = lock, curr = pkg })
+    end
+  end
+  return conflicts
+end
+
+--- @nodoc
+--- @alias Operation
+--- | '"install"'
+--- | '"update"'
+--- | '"remove"'
+--- | '"sync"'
+---
+--- Boilerplate around operations (autocmds, counter initialization, etc.)
+--- @param op Operation
+--- @param fn function
+--- @param pkgs vim.pack.Package[]
+local function exe_op(op, fn, pkgs)
+  if vim.tbl_isempty(pkgs) then
+    vim.notify('Pack: Nothing to ' .. op)
+
+    vim.api.nvim_exec_autocmds('User', {
+      pattern = 'PackDone' .. op:gsub('^%l', string.upper),
+    })
+    return
+  end
+
+  local function after(ok, err, nop)
+    local summary = 'Pack: %s complete. %d ok; %d errors;' .. (nop > 0 and ' %d no-ops' or '')
+    vim.notify(string.format(summary, op, ok, err, nop))
+    vim.cmd('silent! helptags ALL')
+
+    if #BuildQueue ~= 0 then
+      process_build_queue()
+    end
+
+    vim.api.nvim_exec_autocmds('User', { pattern = 'PackDone' .. op:gsub('^%l', string.upper) })
+
+    -- This makes the logfile reload if there were changes while the job was running
+    vim.cmd('silent! checktime ' .. vim.fn.fnameescape(Path.log))
+
+    lock_write()
+  end
+
+  local counter = new_counter(#pkgs, after)
+
+  for _, pkg in ipairs(pkgs) do
+    fn(pkg, counter)
+  end
+end
+
+---@param opts vim.pack.Opts? When omitted or `nil`, retrieve the current
+---       configuration. Otherwise, a configuration table (see |vim.pack.Opts|).
+---@return vim.pack.Opts? : Current pack config if {opts} is omitted.
+function M.config(opts)
+  vim.validate('opts', opts, 'table', true)
+
+  if not opts then
+    return vim.deepcopy(Config, true)
+  end
+
+  vim.iter(opts):each(function(k, v)
+    Config[k] = v
+  end)
+end
+
+--- Register one or more plugins to be installed (see [pack.PackageSpec]())
+---
+--- Example:
+---
+--- ```lua
+--- -- pack will update by itself
+--- pack.register({
+---   "neovim/nvim-lspconfig",
+---   { 'nvim-treesitter/nvim-treesitter', build = ':TSUpdate' },
+---   -- don't load this plugin when registering. The user needs to call `packadd`.
+---   { 'tpope/vim-fugitive', opt = false },
```

Comment:
```suggestion
---   { 'tpope/vim-fugitive', opt = false },
```

If this comment is correct that `opt = false` means the user needs to call `packadd`, this is backwards and needs to change. See [my longer comment here](https://github.com/neovim/neovim/pull/34223#issuecomment-2965125904)

Reply:
Edit... Actually, it seems the comment might be incorrect and thus the generated docs are also incorrect.

Code says this. Looks like it loads at startup if `opt = true` to me!

Although it does it via `packadd` at startup rather than just downloading to the `pack/*/start` directory which I am not the biggest fan of. Sorta ok though still, it will have the same behavior as long as it is called during execution of `init.lua` or `plugin/*` files. Results in a longer rtp though.

```lua
      function(acc, pkg)
        acc[pkg.name] = pkg
        if not pkg.opt then
          pcall(vim.cmd.packadd, pkg.name)
          -- Remove opt from the schema
          ---@diagnostic disable-next-line: inject-field, no-unknown
          pkg.opt = nil
        end
        return acc
      end
```